[perf] improve performance of ReadLogs

xyz-li opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
 Reduce write events from fsnotify.Watcher to reduce calling function isContainerRunning.

When container output logs quickly, and user run command `kubectl logs CONTAINER_ID -f`, then there will be too much function call of `isContainerRunning`. This will consume CPU time of kubelet and containerd.

log generation code
```
package main

import (
     "fmt"
     "time"
     "flag"
)

var dur = flag.Duration("duration", 1 * time.Microsecond, "duration")

func main() {
     i := 0
     flag.Parse()
     for range time.Tick(*dur) {
             fmt.Printf("print log to stand out %d\n", i)
             i++
     }
}
```
1. kubelet pprof with 1000 lines of log per second. ReadLogs takes 19.8% CPU time.
![image](https://github.com/user-attachments/assets/0119fd2d-771b-4721-adee-c6bc0da44734)
2. containerd pprof with 1000 lines of log per second. ContainerStatus_handler takes 20.5% CPU time.
![image](https://github.com/user-attachments/assets/59cdfccf-34d4-40b1-a9e2-550371039df7)
3. kubelet pprof with 100 lines of log per second. ReadLogs takes 14.5% CPU time.
![image](https://github.com/user-attachments/assets/9322ac55-329c-4bee-96b4-c29532475a4a)
4. containerd pprof with 100 lines of log per second. ContainerStatus_handler takes 20.5% CPU time.
![image](https://github.com/user-attachments/assets/abfe340b-062e-4d53-b482-58068b47c91c)
5. kubelet pprof with 1000 lines of log per second. ReadLogs takes 7.7% CPU time.  Write event deduplicated.
![image](https://github.com/user-attachments/assets/9790ccce-ee55-43ec-80fc-d15112896019)
6. containerd pprof with 1000 lines of log per second. ContainerStatus_handler takes 5% CPU time.  Write event deduplicated.
![image](https://github.com/user-attachments/assets/d7bd92b0-820a-4db1-bc9d-d17a7179c645)
7. kubelet pprof with 100 lines of log per second. ReadLogs takes 14.5% CPU time.  Write event deduplicated.
![image](https://github.com/user-attachments/assets/2c6397d3-a23c-4362-8ab4-623952b79cc3)
8. containerd pprof with 100 lines of log per second. ContainerStatus_handler takes 25% CPU time.  Write event deduplicated.
![image](https://github.com/user-attachments/assets/17c98e96-c194-4e0e-a4e8-517041e015d5)


#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```
/area kubelet

xyz-li commented:
/cc @dims 
/cc @saschagrunert

ffromani commented:
/ok-to-test

xyz-li commented:
/retest-required

xyz-li commented:
/retest-required

saschagrunert reviewed (commented):
Code LGTM, this should be considered for v1.32.

saschagrunert commented:
/kind cleanup
/triage accepted

saschagrunert commented:
/release-note-none

xyz-li commented:
/milestone v1.32

xyz-li commented:
kindly ping @saschagrunert

saschagrunert commented:
@xyz-li do you mind giving this another rebase?

xyz-li commented:
> @xyz-li do you mind giving this another rebase?

Done.

bart0sh commented on staging/src/k8s.io/cri-client/pkg/logs/logs.go:
```diff
@@ -306,7 +306,10 @@ func ReadLogs(ctx context.Context, logger *klog.Logger, path, containerID string
 		return fmt.Errorf("failed to open log file %q: %v", path, err)
 	}
-	defer f.Close()
+	defer func() {
+		//nolint:errcheck
+		f.Close()
+	}()
```
```suggestion
	defer f.Close() // nolint:errcheck
```

xyz-li replied:
It may cause resource leaks and close the file twice.
```
func closeFileTwice() {
	fn := "/tmp/abc.txt"
	f, _ := os.Open(fn)
	defer f.Close()

	for i := 0; i < 10; i++ {
		newF, _ := os.Open(fn)
		defer func(f *os.File) {
			if err := f.Close(); err != nil {
				panic(err)
			}
		}(newF)

		f.Close()
		f = newF
	}
}

```

bart0sh replied:
I'm not sure I follow. Your variant will do the same, no?

xyz-li replied:
```
defer f.Close() // nolint:errcheck
// other codes

for{
    // other codes
    defer newF.Close()
}
```
f.Close() executs will be the actual f evaluated at the beginning of the function. Which may cause the file close two times.

```
defer func() {
		//nolint:errcheck
		f.Close()
}()

// other codes
for {
     // other codes
     f.Close()
     f = newF
     r = bufio.NewReader(f)
}
```
 `f = newF` update variable f in the closure, so `f.Close()` always close the latest file and only once.

bart0sh replied:
I still don't see the difference. At least [your example](https://github.com/kubernetes/kubernetes/pull/126385#discussion_r1901823468) and [this code](https://go.dev/play/p/7gWLZAr2sd-) produce the same output.

Can you provide an example to demonstrate the difference between `defer f.Close() // nolint:errcheck` and 
```
defer func() {
		//nolint:errcheck
		f.Close()
}()
``` 
?

bart0sh commented:
> containerd pprof with 100 lines of log per second. ContainerStatus_handler takes 20.5% CPU time.
> containerd pprof with 100 lines of log per second. ContainerStatus_handler takes 25% CPU time. Write event deduplicated.

Can you explain why the fix caused increase in CPU load?

xyz-li commented on staging/src/k8s.io/cri-client/pkg/logs/logs.go:
```diff
@@ -455,33 +467,122 @@ func isContainerRunning(ctx context.Context, logger *klog.Logger, id string, r i
+				case fsnotify.Write:
+					// If there is one Write event in the channel, we can discard this one.
+					// If there is one Create event in the channel, we will reopen the log file, and read from the
+					// start. It's OK to discard the new Write event.
```
@bart0sh  https://github.com/kubernetes/kubernetes/pull/126385#issuecomment-2692587958


When using the `-f` parameter, after reaching the end of the file, every new write to log file triggers `waitLogs`. Inside `waitLogs`, a request to containerd is made to retrieve the `ContainerStatus`. The faster the writes occur, the more frequently waitLogs is called, leading to increased CPU usage by containerd

https://github.com/kubernetes/kubernetes/blob/4806519423456834223934af3ab01d0b1aeeaa46/staging/src/k8s.io/cri-client/pkg/logs/logs.go#L369

https://github.com/kubernetes/kubernetes/blob/4806519423456834223934af3ab01d0b1aeeaa46/staging/src/k8s.io/cri-client/pkg/logs/logs.go#L458-L460


This patch deduplicates the `fsnotify.Write` events from the system, reducing the number of waitLogs calls. As a result, it lowers the CPU usage of both kubelet and containerd

bart0sh replied:
Sorry for not being clear enough. Quoted pprof statistics suggests that CPU time for `ContainerStatus_handler` increased from 20.5% to 25% with this change. According to what you're saying it should decrease, right?

xyz-li replied:
Good question.
A possible reason is that the usage of ContainerStatus_handler has decreased, while the load on other parts of the system has also been reduced. For example, if ContainerStatus_handler originally consumed 2 seconds of CPU time and other parts of containerd consumed 3 seconds of CPU time, then the proportion of ContainerStatus_handler would be 40%. If the usage of ContainerStatus_handler drops to 1 second and the usage of other parts of containerd also drops to 1 second, then the proportion of ContainerStatus_handler would increase to 50%. This might not be a very convincing explanation, but the overall CPU usage of containerd has decreased from 1.32 seconds to 0.72 seconds.

Today, I tested with containerd 2.0.2 on my new node (Intel(R) Core(TM) i3-9100F CPU @ 3.60GHz/16GB), and the ContainerStatus_handler usage decreased from 36.9% to 13.5%. The overall CPU usage also decreased from 5.93 seconds to 1.33 seconds. 

containerd pprof with 100 lines of log per second. ContainerStatus_handler takes 36.9% CPU time.
![image](https://github.com/user-attachments/assets/dc392bda-50fd-43a2-8827-daf25178bfd8)

containerd pprof with 100 lines of log per second. ContainerStatus_handler takes 13.5% CPU time. Write event deduplicated.
![image](https://github.com/user-attachments/assets/37216f8f-dd2d-4f2c-8a7a-19dffbef4b49)

bart0sh commented on staging/src/k8s.io/cri-client/pkg/logs/logs.go:
```diff
@@ -455,33 +467,122 @@ func isContainerRunning(ctx context.Context, logger *klog.Logger, id string, r i
+			<-ticker.C
+		}()
+
+		errRetry := 2
```
Why the number of retries decreased from 5 to 2?

bart0sh commented on staging/src/k8s.io/cri-client/pkg/logs/logs_test.go:
```diff
@@ -596,3 +596,61 @@ func TestOnlyStdoutStream(t *testing.T) {
+
+			w := &fsnotify.Watcher{Events: make(chan fsnotify.Event), Errors: make(chan error)}
+			dedup := newDedupWriteEvents(w, "0.log", nil)
+			dedup.dedupEvents(context.Background())
```
I'd suggest to use ktesting.Init(t) to get a text context in the beginning of the test and then reuse it in the loop.

xyz-li replied:
Sorry,  I can not understand what you mean.

bart0sh replied:
Sorry for being unclear. I suggested to have `tCtx := ktesting.Init(t)` in the beginning of the test and then use it here instead of `context.Background()` this way: `dedup.dedupEvents(tCtx)`. This is a common approach in k/k unit tests, for example https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet_test.go#L3361

xyz-li replied:
I've removed the `ctx` parameter. It's no longer necessary.

xyz-li commented on staging/src/k8s.io/cri-client/pkg/logs/logs.go:
```diff
@@ -455,33 +467,122 @@ func isContainerRunning(ctx context.Context, logger *klog.Logger, id string, r i
+			case <-ctx.Done():
+				return
+
+			case err := <-de.w.Errors:
```
I have rewritten it to minimize the modifications to the function waitLogs.

bart0sh replied:
It still looks complicated to me. The code introduces a proxy around `fsnotify.Watcher`, runs it as a goroutine and copies `fsnotify.Watcher` events to its own channels just to skip some Write events. Just wondering if we can do the same without the proxy, in the `waitLogs` function?

xyz-li commented:
/retest-required

xyz-li commented:
/test pull-kubernetes-unit

bart0sh commented on staging/src/k8s.io/cri-client/pkg/logs/dedup.go:
```diff
@@ -0,0 +1,102 @@
+	}
+}
+
+type dedupWriteEvents struct {
+	w           *fsnotify.Watcher
+	logFileName string
+
+	// Events sends fsnotify events.
+	Events chan fsnotify.Event
+	// Errors sends fsnotify errors.
+	Errors chan error
+}
```
nit: Can you move this up so the structure is defined before its methods?

xyz-li commented on staging/src/k8s.io/cri-client/pkg/logs/logs.go:
```diff
@@ -356,17 +360,23 @@ func ReadLogs(ctx context.Context, logger *klog.Logger, path, containerID string
 					}
+
+					dedupWatcher = newDedupWriteEvents(baseName, watcher)
+					go dedupWatcher.dedupLoop()
```
We need this goroutine to read events and discard the duplicated events. If we don't consume these events, an `ErrEventOverflow` error will occur in the `Errors` channel.

xyz-li commented:
/retest-required

bart0sh commented:
/priority backlog

bart0sh commented on staging/src/k8s.io/cri-client/pkg/logs/logs.go:
```diff
@@ -455,7 +464,7 @@ func isContainerRunning(ctx context.Context, logger *klog.Logger, id string, r i
-func waitLogs(ctx context.Context, logger *klog.Logger, id string, logName string, w *fsnotify.Watcher, runtimeService internalapi.RuntimeService) (bool, bool, error) {
+func waitLogs(ctx context.Context, logger *klog.Logger, id string, logName string, w *dedupWriteEventsWatcher, runtimeService internalapi.RuntimeService) (bool, bool, error) {
 	// no need to wait if the pod is not running
 	if running, err := isContainerRunning(ctx, logger, id, runtimeService); !running {
```
Just wondering if we can get the same effect not checking if container is running for `waitDuration` ?

xyz-li replied:
I believe it is feasible, provided that one can endure the waitDuration seconds after the container's termination。

xyz-li replied:
3. profile of this commit. 100 lines of log per second
<img width="1439" alt="image" src="https://github.com/user-attachments/assets/4f0b2ab5-ac80-4f2e-9caf-7f73bee0b7fa" />
4. profile of https://github.com/xyz-li/kubernetes/commit/0d34a8af6ce85d3b13cc5fdff92906bf18fa2896 100 lines of log per second
<img width="1503" alt="image" src="https://github.com/user-attachments/assets/895e0c39-266f-4732-b4ec-4761651e530c" />

bart0sh replied:
Does the code in https://github.com/xyz-li/kubernetes/commit/0d34a8af6ce85d3b13cc5fdff92906bf18fa2896 skip any `Write` events?

xyz-li replied:
https://github.com/xyz-li/kubernetes/commit/0d34a8af6ce85d3b13cc5fdff92906bf18fa2896 just reduces calls to `isContainerRunning`.

xyz-li replied:
1.  profile of  https://github.com/xyz-li/kubernetes/commit/81d18e153f4e4ec530586ab9e8b67736e73adeae  1000 lines of log per second

<img width="1512" alt="image" src="https://github.com/user-attachments/assets/27532232-841a-49a8-aff2-4f0394ee9f7e" />

2. profile of this commit. 1000 lines of log per second. Can't find `dedupLoop` in the profile.
<img width="1509" alt="image" src="https://github.com/user-attachments/assets/de716d16-15ae-4148-bb11-792d60d147b8" />

3.  profile of  https://github.com/xyz-li/kubernetes/commit/81d18e153f4e4ec530586ab9e8b67736e73adeae  100 lines of log per second
<img width="1512" alt="image" src="https://github.com/user-attachments/assets/9d67fcda-debf-4395-8ad1-850f8710ffdc" />

4. profile of this commit. 100 lines of log per second. `dedupLoop` consumes  about 1.4% cpu time.
<img width="1506" alt="image" src="https://github.com/user-attachments/assets/bfd5c2ef-08bb-4d6e-ab1c-9bd8479a5873" />

5. All the profiles.
[profile.tgz](https://github.com/user-attachments/files/19451398/profile.tgz)

bart0sh replied:
Can you provide some numbers for CPU load so we can decide which way to go?

xyz-li replied:
CPU: Intel(R) Core(TM) i3-9100F CPU @ 3.60GH
Memory: 16GB
OS: Ubuntu 24.04.1 LTS

profile time: 30s

1.  profile of https://github.com/xyz-li/kubernetes/commit/aacff57f767d99ad1367b2e1fc2f838a6c0fe334 1000 lines of log per second

<img width="1512" alt="image" src="https://github.com/user-attachments/assets/27532232-841a-49a8-aff2-4f0394ee9f7e" />
<img width="1499" alt="image" src="https://github.com/user-attachments/assets/7a5cf2de-3f3c-46fa-bc0b-c6dae017f208" />


2. profile of this commit. 1000 lines of log per second. Can't find `dedupLoop` in the profile.
<img width="1509" alt="image" src="https://github.com/user-attachments/assets/de716d16-15ae-4148-bb11-792d60d147b8" />
<img width="1507" alt="image" src="https://github.com/user-attachments/assets/192abefc-cb74-4079-93b5-0dee2e3bc5f4" />


3.  profile of https://github.com/xyz-li/kubernetes/commit/aacff57f767d99ad1367b2e1fc2f838a6c0fe334 100 lines of log per second
<img width="1512" alt="image" src="https://github.com/user-attachments/assets/9d67fcda-debf-4395-8ad1-850f8710ffdc" />
<img width="1508" alt="image" src="https://github.com/user-attachments/assets/617beeaf-5017-48dd-aee4-ce5477d8027a" />


4. profile of this commit. 100 lines of log per second. `dedupLoop` consumes  about 1.4% cpu time.
<img width="1506" alt="image" src="https://github.com/user-attachments/assets/bfd5c2ef-08bb-4d6e-ab1c-9bd8479a5873" />
<img width="1507" alt="image" src="https://github.com/user-attachments/assets/a4f46f7b-3a51-4601-a2fc-f04a66f7e6bd" />


5. All the profiles.
[profile.tgz](https://github.com/user-attachments/files/19459455/profile.tgz)

bart0sh replied:
How much ReadLogs consumes in total for 2 implementation we're comparing: the one with dedupLoop and the one where dedup logic is done by waitLogs?

xyz-li replied:
| parameter      | this commit | aacff57|
| ----------- | ----------- |-------|
| 1000 lines/s      | ReadLogs: 250ms  total: 6.81s     | ReadLogs: 640ms  total: 6.64s|
| 500 lines/s     | ReadLogs: 210ms total: 4.52s         |ReadLogs: 330ms  total: 4.48s|
| 200 lines/s   | ReadLogs: 150ms total: 2.72s            |ReadLogs: 180ms  total: 2.68s|
| 100 lines/s   | ReadLogs: 70ms total: 2.06s          |ReadLogs: 80ms  total: 2.3s|


I ran profiling with  1000, 500, 200, and 100 lines per second but saw no significant CPU usage for `dedupLoop`. It likely consumes minimal CPU time.

**In commit aacff57, As log output rates increase, go channel race conditions consume substantial CPU resources.**

1000 lines/s 
<img width="1509" alt="image" src="https://github.com/user-attachments/assets/bd499996-9c61-4d28-b691-98dbd8d0d106" />
500 lines/s 
<img width="1507" alt="image" src="https://github.com/user-attachments/assets/6b694903-1a14-4d09-bda6-47a9d15033c1" />





aacff57:
100 lines/s 
<img width="1511" alt="image" src="https://github.com/user-attachments/assets/f70a5ed6-147a-4a24-8ffb-81302b198413" />
200 lines/s 
<img width="1511" alt="image" src="https://github.com/user-attachments/assets/26c5e7ab-a96c-4414-b338-89d1a0df01e0" />
500 lines/s 
<img width="1510" alt="image" src="https://github.com/user-attachments/assets/a635487c-f7b0-4934-bcac-aed7a44aed05" />
1000 lines/s 
<img width="1498" alt="image" src="https://github.com/user-attachments/assets/fb680c4d-c91d-4dbe-8965-524e21ead796" />

this commit:
100 lines/s 
<img width="1509" alt="image" src="https://github.com/user-attachments/assets/22a06742-b3a1-47cd-b0bd-62f556aa9140" />

200 lines/s
<img width="1512" alt="image" src="https://github.com/user-attachments/assets/193810c0-c0a5-41ea-be98-ec322e51dd50" />

500 lines/s
<img width="1511" alt="image" src="https://github.com/user-attachments/assets/36f6e3ee-a0ee-430b-b41a-a2a85f3bfa16" />

1000 lines/s
<img width="1507" alt="image" src="https://github.com/user-attachments/assets/5c248958-cf3e-4c4b-95cf-8eb906d1ba78" />





[profile.tgz](https://github.com/user-attachments/files/19483422/profile.tgz)

bart0sh replied:
`Total` numbers look quite similar for both implementations. I'd vote for the one without `dedupWriteEventsWatcher`.

xyz-li committed:
[perf] reduce checks of container status && fsnotify.Write event

When container output logs quickly, and user run command kubectl logs CONTAINER_ID -f,
then there will be too much function call of isContainerRunning.
This will consume CPU time of kubelet and containerd.

Signed-off-by: xyz-li <hui0787411@163.com>

xyz-li commented:
/retest-required

bart0sh commented on staging/src/k8s.io/cri-client/pkg/logs/logs.go:
```diff
@@ -452,30 +456,76 @@ func isContainerRunning(ctx context.Context, logger *klog.Logger, id string, r i
+				if time.Since(wl.lastWriteEvent) > waitDuration {
+					wl.lastWriteEvent = time.Now()
+					return true, false, nil
+				}
```
Can you explain how it's guaranteed that even if we skip some write events, eventually all file content will be fully read, especially in the case when file is rotated?

bart0sh replied:
Thank you for the explanations!
What happens if in the case 1 file is deleted while we're waiting and we skipped a lot of events. Will the log content that was produced while we're waiting be lost?

So, in case of rotation, old file will still be read even though it's renamed?

Are these cases covered by the tests?

xyz-li replied:
As long as a file descriptor remains open, the system will preserve the file from deletion. The application may continue reading the file through this descriptor until it is closed.

This test covers these case.
https://github.com/kubernetes/kubernetes/blob/da61f24acfb54bb8d382d2875ade8fca6af02f79/staging/src/k8s.io/cri-client/pkg/logs/logs_test.go#L215

xyz-li commented:
/retest-required

xyz-li commented:
Kindly ping @SergeyKanzhelev @mrunalp

SergeyKanzhelev commented on staging/src/k8s.io/cri-client/pkg/logs/logs.go:
```diff
@@ -452,30 +456,84 @@ func isContainerRunning(ctx context.Context, logger *klog.Logger, id string, r i
+	}
+
+	if wl.recreated {
+		wl.recreated = false
+		// Reopen the log file now.
+		return true, true, nil
```
really hard to understand this. recreated is set to false and then we return recreated as `true`

SergeyKanzhelev commented on staging/src/k8s.io/cri-client/pkg/logs/logs_test.go:
```diff
@@ -289,7 +289,8 @@ func TestReadRotatedLog(t *testing.T) {
 
-	time.Sleep(20 * time.Millisecond)
+	// waitLog.wait may wait 1 second
+	time.Sleep(1500 * time.Millisecond)
```
should we test both - with a second wait and without 1 second wait? So both codepath will be tested?

xyz-li committed:
variable renaming && test

Signed-off-by: xyz-li <hui0787411@163.com>

xyz-li commented on staging/src/k8s.io/cri-client/pkg/logs/logs.go:
```diff
@@ -378,7 +383,6 @@ func ReadLogs(ctx context.Context, logger *klog.Logger, path, containerID string
 						}
 						return fmt.Errorf("failed to open log file %q: %v", path, err)
 					}
-					defer newF.Close()
```
https://go.dev/play/p/gb1wLLQ6u_K


```
defer f.Close() // nolint:errcheck
// other codes

for{
    // other codes
    defer newF.Close()
    f.Close()
    f = newF
}
```
f.Close() will be the actual f evaluated at the beginning of the function. Which may cause the file close two times.

```
defer func() {
		//nolint:errcheck
		f.Close()
}()

// other codes
for {
     // other codes
     f.Close()
     f = newF
     r = bufio.NewReader(f)
}
```
 `f = newF` update variable f in the closure, so `f.Close()` always close the latest file and only once.

xyz-li replied:
https://github.com/kubernetes/kubernetes/pull/124298

I previously created a PR to fix this issue, it has been closed by me.

xyz-li commented:
/retest-required
